src/core/network/polling.ts

```diff
@@ -0,0 +1,61 @@
+import * as tt from '../../telegram-types'
+import AbortController from 'abort-controller'
+import ApiClient from './client'
+import d from 'debug'
+import { promisify } from 'util'
+const wait = promisify(setTimeout)
+const debug = d('telegraf:polling')
+
+export class Polling {
+  private readonly abortController = new AbortController()
+  private offset = 0
+  constructor(
+    private readonly telegram: ApiClient,
+    private readonly allowedUpdates: readonly tt.UpdateType[]
+  ) {}
+
+  async *[Symbol.asyncIterator](): AsyncGenerator<tt.Update[], void, void> {
+    debug('Starting long polling')
+    do {
+      try {
+        const updates = await this.telegram.callApi(
+          'getUpdates',
+          {
+            timeout: 50,
+            offset: this.offset,
+            allowed_updates: this.allowedUpdates,
+          },
+          this.abortController
+        )
+        const last = updates[updates.length - 1]
+        if (last !== undefined) {
+          this.offset = last.update_id + 1
+        }
+        yield updates
+      } catch (err) {
+        if (err.name === 'AbortError') return
+        if (err.code === 401 || err.code === 409) throw err
+        const retryAfter: number = err.parameters?.retry_after ?? 5
+        debug('Failed to fetch updates, retrying after %ds.', retryAfter, err)
+        await wait(retryAfter * 1000)
+      }
+    } while (!this.abortController.signal.aborted)
+    await this.confirmUpdates()
+  }
+
+  async confirmUpdates() {
+    debug('Confirming updates...')
+    await this.telegram.callApi('getUpdates', { offset: this.offset, limit: 1 })
+  }
```

Comment:
It is not clear why we need this call. Please add a brief comment explaining why it is necessary to fetch an update and immediately discard it.

Also, it seems to be called inconsistently depending on where an error is thrown, but I cannot really pinpoint this without knowing what it actually does.

Is it relevant here to pass `allowed_updates`?

Reply:
I could link to https://github.com/telegraf/telegraf/issues/851#issuecomment-736025893 in a comment...

Or, I could do something that sounds more appealing the more I think about it: call `confirmUpdates` in `finally` clause in `loop`. Disadvantage: call would happen even if not necessary. Advantages:

+ Bulletproof. It'll always be called, always exactly once (unless the process is killed of course).
+ There's no need to call `confirmUpdates` anywhere else, so it can be made `private`.


> Is it relevant here to pass `allowed_updates`?

Not really. We don't care what we get. And it defaults to value from previous call anyway.

Reply:
> I could link to https://github.com/telegraf/telegraf/issues/851#issuecomment-736025893 in a comment...

Yes, or just explain it :)

I won't be the last one to wonder about this, and a quick sentence or two is much more convenient.

> call `confirmUpdates` in `finally` clause in `loop`

That actually sounds much clearer. I think reducing the time from dozens of seconds to a few msec justifies one superfluous call every blue moon, just to be safe.

Reply:
I'm still open to better `debug` message

Reply:
I'll throw some suggestions at you. Just pick any, or none, or any combination:
- `Saving update offset`
- `Ensuring correct update offset`
- `Acknowledging updates`
- `Confirming received updates to Telegram`
- `Syncing update queue`
- `Securing update index`
I can keep going if you dislike all of them.